core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala

```diff
@@ -80,8 +87,58 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
   //periodically emit metrics (don't need to do this for each message!)
   context.system.scheduler.schedule(30.seconds, 10.seconds, self, EmitMetrics)
 
+  // Key is ColdStartKey, value is the number of cold Start in minute
+  var coldStartCount = immutable.Map.empty[ColdStartKey, Int]
+
   backfillPrewarms(true)
 
+  // check periodically every 1 minute, adjust prewarmed container(delete if unused for some time and create some increment containers)
+  context.system.scheduler.schedule(1.minute, 1.minute, self, AdjustPrewarmedContainer)
+
+  def adjustPrewarmedContainer(): Unit = {
```

Comment:
Can you DRY up the code in `adjustPrewarmedContainer` and `backfillPrewarms`, possibly using the same function for both?

Reply:
hm.. it seems `adjustPrewarmedContainer`'s logic is not same as `backfillPrewarms`, and in its inner,  it seems can't extract the common code as well.

Reply:
There is a lot of duplicated delicate logic - these need to be consolidated, I think.

Reply:
Here is a proposed version that:
- consolidates `backfillPrewarms()` and `adjustPrewarmedContainer()`
- removes `prewarmContainerIfpossible()` 
```
  def adjustPrewarmedContainer(init: Boolean, scheduled: Boolean): Unit = {
    prewarmConfig.foreach { config =>
      // Delete unused prewarmed container until minCount is reached
      val kind = config.exec.kind
      val memory = config.memoryLimit

      val runningCount = prewarmedPool.count {
        //done starting, and not expired
        case (_, p @ PreWarmedData(_, `kind`, `memory`, _, _)) if !p.isExpired() => true
        //started but not finished starting (or expired)
        case _ => false
      }
      val startingCount = prewarmStartingPool.count(p => p._2._1 == kind && p._2._2 == memory)
      val currentCount = runningCount + startingCount
      //determine how many are needed
      val desiredCount: Int =
        if (init) config.initialCount
        else {
          if (scheduled) {
            //scheduled/reactive config backfill
            config.reactive
              .map(c => getReactiveCold(c, kind, memory).getOrElse(c.minCount)) //reactive -> desired is either cold start driven, or minCount
              .getOrElse(config.initialCount) //not reactive -> desired is always initial count
          } else {
            //normal backfill after removal - make sure at least minCount or initialCount is started
            config.reactive.map(_.minCount).getOrElse(config.initialCount)
          }
        }
      if (currentCount < desiredCount) {
        logging.info(
          this,
          s"found ${currentCount} started and ${startingCount} starting; ${if (init) "initing" else "backfilling"} ${desiredCount - currentCount} pre-warms to desired count: ${desiredCount} for kind:${config.exec.kind} mem:${config.memoryLimit.toString}")(
          TransactionId.invokerWarmup)
        (currentCount until desiredCount).foreach { _ =>
          prewarmContainer(config.exec, config.memoryLimit, config.reactive.map(_.ttl))
        }

      }

      //remove expired
      config.reactive.foreach { reactiveValue =>
        prewarmedPool
          .filter { warmInfo =>
            warmInfo match {
              case (_, p @ PreWarmedData(_, `kind`, `memory`, _, _)) if p.isExpired() => true
              case _                                                                  => false
            }
          }
          .drop(reactiveValue.minCount) //keep minCount even if expired
          .map(_._1 ! Remove)
      }
    }
    if (scheduled) {
      //clear coldStartCounts each time scheduled event is processed to reset counts
      coldStartCount = immutable.Map.empty[ColdStartKey, Int]
    }
  }

  def getReactiveCold(config: ReactivePrewarmingConfig, kind: String, memory: ByteSize): Option[Int] = {
    coldStartCount.get(ColdStartKey(kind, memory)).map { value =>
      // Let's assume that threshold is `2`, increment is `1` in runtime.json
      // if cold start number in previous minute is `2`, requireCount is `2/2 * 1 = 1`
      // if cold start number in previous minute is `4`, requireCount is `4/2 * 1 = 2`
      math.min(math.max(config.minCount, (value / config.threshold) * config.increment), config.maxCount)
    }
  }
```

Reply:
Great! codes become more clear.
Updated accordingly.